Updates thresholding

marieindilewitsch opened the pull request:
1)Changes in constants: adjusted kernel size to (5,5) for smoother contours (did not change the output), changes contour level to 4 to get a higher resolution (did not change the output)

2)Added logging info in segmentation request processing, which doesn't show when being executed,

3)In utils whole slide, I adjusted the background mask generation

4)In utils segment image I adjusted the parameters in adaptive Threshold and added the Otsu_threshold function to allow for better distinguishing between fore- and background.

marieindilewitsch committed:
Updates thresholding 

1)Changes in constants: adjusted kernel size to (5,5) for smoother contours (did not change the output), changes contour level to 4 to get a higher resolution (did not change the output)

2)Added logging info in segmentation request processing, which doesn't show when being executed,

3)In utils whole slide, I adjusted the background mask generation

4)In utils segment image I adjusted the parameters in adaptive Threshold and added the Otsu_threshold function to allow for better distinguishing between fore- and background.

josh-freeman commented on backend/segmentation_service.py:
```diff
@@ -27,6 +27,7 @@
     tile_and_save_contours,
 
 )
+np.random.seed(42)
```
There are two problems with this
- A random seed should not be set at the top of a file like this. Please find the main() in which this needs to be set and set it there.
- Add a RANDOM_SEED constant and import it

marieindilewitsch replied:
Random seed is added now in the constants.py and imported in the wsi.ipynb from where the experiments are run.

josh-freeman commented on backend/segmentation_service.py:
```diff
@@ -36,7 +37,7 @@ def process_image_segmentation_request(file: io.BytesIO) -> dict:
     gray_array, image = preprocess_image(file)
     original_image = np.array(image)
-
+    print("hereeee")
```
do not commit this

marieindilewitsch replied:
I deleted this.

josh-freeman commented on backend/utils_segment_image.py:
```diff
@@ -11,6 +11,7 @@
     RPB_THRESHOLD_PERCENTILE,
     GRABCUT_ITERATIONS
 )
+np.random.seed(42)
```
There are two problems with this

    A random seed should not be set at the top of a file like this. Please find the main() in which this needs to be set and set it there.
    Add a RANDOM_SEED constant and import it

marieindilewitsch replied:
Same here, deleted this line and imported where it was needed.

josh-freeman commented on backend/utils_segment_image.py:
```diff
@@ -63,8 +64,8 @@ def get_background_mask(image: np.ndarray) -> np.ndarray:
 def get_nuclei_mask(image: np.ndarray) -> np.ndarray:
-    thresh = cv2.adaptiveThreshold(image, 255, cv2.ADAPTIVE_THRESH_MEAN_C, 
-                                   cv2.THRESH_BINARY_INV, 11, 1)
+    thresh = cv2.adaptiveThreshold(image, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, 
```
Is this change needed? If so, justify it in the commit message. If it is part of an experiment, then add the two possible choices to a list in constants.py.

marieindilewitsch replied:
Using the Gaussian Adaptive Threshold is leads to a better significance then just the Mean Adaptive Threshold (p-value for all samples improved from 0.042 with Mean to 0.038 with Gaussian), will specify this in the next commit of my Pull request.

josh-freeman replied:
I am not sure what you mean here by specifying in the next commit, but I will add this to the commit that this PR is squashed into.

josh-freeman commented on backend/utils_segment_image.py:
```diff
@@ -63,8 +64,8 @@ def get_background_mask(image: np.ndarray) -> np.ndarray:
-    thresh = cv2.adaptiveThreshold(image, 255, cv2.ADAPTIVE_THRESH_MEAN_C, 
-                                   cv2.THRESH_BINARY_INV, 11, 1)
+    thresh = cv2.adaptiveThreshold(image, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, 
+                                   cv2.THRESH_BINARY_INV, 13, 1)
```
Magic number. This 13 should be added to constants.py and imported.

marieindilewitsch replied:
Done.

josh-freeman commented on backend/utils_segment_whole_slide.py:
```diff
@@ -11,17 +11,26 @@
     # Apply a kernel to smoothen the mask
-    kernel = np.ones((3, 3), np.uint8)
-    smoothed_mask = cv2.morphologyEx(mask_background, cv2.MORPH_CLOSE, kernel)
+    kernel = np.ones((5, 5), np.uint8)
```
magic number. should be from constants.py

marieindilewitsch replied:
Done.

josh-freeman commented on backend/utils_segment_whole_slide.py:
```diff
@@ -11,17 +11,26 @@
+    gray = cv2.cvtColor(level_array, cv2.COLOR_RGB2GRAY)
 
+    # Smooth the grayscale image
+    blurred = cv2.GaussianBlur(gray, (9, 9), 0)
```
magic number. should be from constants.py

marieindilewitsch replied:
In the updated code, I deleted this line, and used the previous version of the get_saved_contours function, so this comment is solved.

josh-freeman commented on backend/utils_segment_whole_slide.py:
```diff
@@ -35,7 +44,7 @@ def get_saved_contours(level_array):
     x0: int, y0: int,
     w0: int, h0: int,
-    tile_size: int,
+    tile_size:  TILE_SIZE,
```
Incorrect: `TILE_SIZE` is not a type!
```suggestion
    tile_size:  int = TILE_SIZE,
```

marieindilewitsch replied:
Changed back to int.

josh-freeman commented on backend/utils_segment_whole_slide.py:
```diff
@@ -175,7 +184,7 @@ def process_one_tile(
 def create_mosaic(
     tile_grid: List[List[np.ndarray]],
-    tile_size: int
+    tile_size:  TILE_SIZE
```
Incorrect: TILE_SIZE is not a type!
```suggestion
    tile_size:  int = TILE_SIZE
```

marieindilewitsch replied:
Changed back to int.

josh-freeman commented on backend/utils_segment_whole_slide.py:
```diff
@@ -222,7 +231,7 @@ def tile_and_save_contours(
     saved_contours,
     downsample_factor: float,
-    tile_size: int = 512,
+    tile_size:  TILE_SIZE,
```
Incorrect: TILE_SIZE is not a type!
```suggestion
    tile_size:  int = TILE_SIZE
```

marieindilewitsch replied:
Changed back to int.

marieindilewitsch committed:
Update constants.py

marieindilewitsch committed:
Update segmentation_service.py

marieindilewitsch committed:
Update utils_segment_image.py

marieindilewitsch committed:
Update utils_segment_whole_slide.py

josh-freeman commented on backend/constants.py:
```diff
@@ -1,6 +1,6 @@
 import os
-INPUT_PATH = "dataset/sample_tiles/imgs/"
-OUTPUT_PATH = "dataset/sample_tiles/pred/"
+#INPUT_PATH = "dataset/sample_tiles/imgs/"
+#OUTPUT_PATH = "dataset/sample_tiles/pred/"
```
Why is this commented out?? Either remove it entirely (and justify in the commits) or *do not commit this*

marieindilewitsch replied:
removed.

josh-freeman commented on backend/segmentation_service.py:
```diff
@@ -18,7 +18,7 @@
     TOP_BIGGEST_CONTOURS_TO_OBSERVE,
     CONTOUR_LEVEL,
-    MIN_FRACTION_OF_TILE_INSIDE_CONTOUR
+    MIN_FRACTION_OF_TILE_INSIDE_CONTOUR,
```
useless comma 
```suggestion
    MIN_FRACTION_OF_TILE_INSIDE_CONTOUR
```

marieindilewitsch replied:
Removed.

josh-freeman commented on backend/segmentation_service.py:
```diff
@@ -27,6 +27,7 @@
     tile_and_save_contours,
 
 )
+
```
useless breakline
```suggestion
```

marieindilewitsch replied:
removed.

josh-freeman commented on backend/segmentation_service.py:
```diff
@@ -101,7 +102,7 @@ def process_ndpi_segmentation_request(
     # Get contours from the array
     saved_contours = get_saved_contours(level_array)
-    
+    print(f"Found {len(saved_contours)} contours")
```
Stay consistent: use loggin, not printing.
```suggestion
    logging.info(f"Found {len(saved_contours)} contours")
```

marieindilewitsch replied:
Changed.

josh-freeman commented on backend/segmentation_service.py:
```diff
@@ -121,13 +122,15 @@ def process_ndpi_segmentation_request(
         logging.info(f"Overlay image with contours saved to: {overlay_image_path}")
 
     # Create tissue/non-tissue mask
+
+    
```
useless breaklines
```suggestion
```

marieindilewitsch replied:
Deleted.

josh-freeman commented on backend/segmentation_service.py:
```diff
@@ -121,13 +122,15 @@ def process_ndpi_segmentation_request(
     mask_background = get_background_mask(level_array)      # 255 = background, 0 = tissue
     tissue_mask = cv2.bitwise_not(mask_background)          # invert: 255 = tissue, 0 = background
-
+    
```
useless breaklines
```suggestion
```

marieindilewitsch replied:
Deleted.

josh-freeman commented on backend/segmentation_service.py:
```diff
@@ -121,13 +122,15 @@ def process_ndpi_segmentation_request(
     saved_contours = get_saved_contours(level_array)
     saved_contours = get_top_biggest_contours(saved_contours, top_n=top_biggest_contours_to_observe)
-
+  
```
Useless breaklines
```suggestion
```

marieindilewitsch replied:
Deleted.

josh-freeman commented on backend/segmentation_service.py:
```diff
@@ -141,4 +144,4 @@ def process_ndpi_segmentation_request(
     )
-    return result
\ No newline at end of file
+    return result
```
What changed here? unclear

marieindilewitsch replied:
Also unclear on that, seems to be no change.

marieindilewitsch replied:
Seemed to be an unnecessary break. Deleted.

josh-freeman commented on backend/utils_segment_whole_slide.py:
```diff
@@ -16,12 +16,16 @@
 import openslide
 logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
 
+
```
useless breakline
```suggestion
```

marieindilewitsch replied:
Deleted.

josh-freeman commented on backend/utils_segment_whole_slide.py:
```diff
@@ -16,12 +16,16 @@
 def get_saved_contours(level_array):    
     mask_background = get_background_mask(level_array)
-
+    
```
useless tab
```suggestion
```

marieindilewitsch replied:
Deleted.

josh-freeman commented on backend/utils_segment_whole_slide.py:
```diff
@@ -16,12 +16,16 @@
     kernel = np.ones((3, 3), np.uint8)
     smoothed_mask = cv2.morphologyEx(mask_background, cv2.MORPH_CLOSE, kernel)
+    
+    #closed = cv2.morphologyEx(binary_mask, cv2.MORPH_CLOSE, kernel)
+    #cleaned = cv2.morphologyEx(closed, cv2.MORPH_OPEN, kernel)
```
why was this necessary? Justfify this.

marieindilewitsch replied:
I removed this, this was a test, if including extra morphological operations would improve the contours, code is deleted now.

josh-freeman commented on backend/utils_segment_whole_slide.py:
```diff
@@ -16,12 +16,16 @@
     # Apply a kernel to smoothen the mask
     kernel = np.ones((3, 3), np.uint8)
     smoothed_mask = cv2.morphologyEx(mask_background, cv2.MORPH_CLOSE, kernel)
+    
```
useless breakline
```suggestion
```

marieindilewitsch replied:
Deleted.

josh-freeman commented on backend/utils_segment_whole_slide.py:
```diff
@@ -35,7 +39,7 @@ def get_saved_contours(level_array):
     x0: int, y0: int,
     w0: int, h0: int,
-    tile_size: int,
+    tile_size:  int,
```
useless double space 
```suggestion
    tile_size: int,
```

marieindilewitsch replied:
Deleted.

josh-freeman commented:
As a general comment, I would say that this is not ready yet for merging. None of my comments have been resolved or answered for those that are questions. As a side note, commit messages are very lackluster. Please read https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/ to get an idea of what commit messages should look like.

marieindilewitsch committed:
Update constants.py

Removed unnecessary break lines and comments, added RANDOM_SEED and THRES_PARAMETER constants. Adjusted the parameters for TILE_SIZE and CONTOUR_LEVEL to get an image with cleaner contours that has higher resuloution.

marieindilewitsch committed:
Update segmentation_service.py

marieindilewitsch committed:
Update utils_segment_image.py

marieindilewitsch committed:
Update utils_segment_whole_slide.py

josh-freeman commented:
Please stick to conventional commits. See https://www.conventionalcommits.org/en/v1.0.0/.
